### ISSUE TITLE
[roadmap] verl development Q2

### ISSUE THREAD

**Issue by eric-haibin-lin** [volcengine/verl#710]
Past roadmap for reference: https://github.com/volcengine/verl/issues/22

## Agentic RL: Environment interaction & tool support [P0]
- [x] Integrate [SandBox](https://github.com/bytedance/SandboxFusion/tree/main) for code generation tasks (more mature than verl's current code sandbox)
- [ ] search / environment interaction via http / grpc 
- [x] multi-turn optimizations, <del>better kv cache management and streaming generations (potential inference engine dependency)</del> https://github.com/volcengine/verl/pull/1037 https://github.com/volcengine/verl/pull/1138  will leave kv cache optimization to inference engines
- [ ] further multi-turn rollout improvements see https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/issues/131 

## Scaling up RL & system performance [P0]
- [ ] Ring Attention
- [x] Ulyssess sequence parallel for VLM models, e.g Qwen2VL
- [x] reference system tuning script for best RL throughput on different types of accelerators
- [x] multi-node rollout (potential inference engine dependency)
- [x] alignment loss fused kernels https://github.com/volcengine/verl/pull/1212

## Usability improvement 
- [x] make the current ray trainer easier to extend (without modifying verl source code or forks). Currently users can define their custom reward via command line without modifying verl source code. Ideally in the main training loop we should allow custom dataset as well https://github.com/volcengine/verl/blob/main/verl/trainer/ppo/ray_trainer.py#L41. 
- [x] collect benchmark result on torchtitan vs megatron
- [ ] support TorchTitan nd parallelism for better usability 

## Latest Model & Algorithm Support
See https://verl.readthedocs.io/en/latest/advance/fsdp_extension.html for adding models with FSDP backend
See https://verl.readthedocs.io/en/latest/advance/megatron_extension.html for adding models with Megatron backend. 
- [ ] gemma3 https://github.com/volcengine/verl/pull/1613/files
- [x] deepseek v3 - since it's large, we should start with SFT for correctness verification, optimize it with fused kernels/recomputation, before moving to RL https://github.com/volcengine/verl/issues/708 https://github.com/volcengine/verl/pull/1771
- [x] qwen3 & qwen3-moe https://github.com/huggingface/transformers/pull/36878 
and any other popular models. 
- [ ] OLMo2
- [x] Dr. GRPO https://github.com/volcengine/verl/issues/742

## Component Continuous Updates
- [x] verify ulysses sequence parallelism support works with latest version of transformers >= v4.50
- [x] replace FSDP1 with FSDP2 https://github.com/volcengine/verl/pull/1026 
- [x] add activation offloading optimization https://github.com/volcengine/verl/pull/1220/files 

## dataset & benchmark
- [ ] gpqa diamond (english)
- [ ] LiveCodeBench (code)
- [ ] SWE-bench Verified (code)
- [ ] CNMO 2024 (math)
- [ ] codecontests (Code Generation)
- [ ] TACO (Code Generation)
- [ ] competition_math (Math)

Please also help provide scripts to reproduce evaluation performance of public released models. 

## Efficient RL / codesign [P1]
- [x] lora support for RL, and provide convergence report  https://github.com/volcengine/verl/pull/1127

## Wide Hardware Coverage
Make the experience on non-nvidia GPUs more smooth
- [x] stable Ascend NPUs suppport, with reproducible examples and logs 
- [x] stable AMD GPUs suppport, with sglang
- [x] AMD GPU with mcore support

## Make verl easier to extend with custom train/infer engine and roles
- [x] https://github.com/volcengine/verl/pull/1424
- [x] https://github.com/volcengine/verl/issues/1371 

# other community requests

- retool https://github.com/volcengine/verl/issues/1169 
- load balance #658

**Comment by eric-haibin-lin** [comment]
> Are there any plans to align the alignment GRPO in Megatron training with FSDP training? We've been struggling for weeks to match the performance(loss/ eval score). We alignment fist step loss and grad (diff less than 0.01), but Megatron's performance degrades after training for several hundred steps.

@ETOgaosion is your Megatron GRPO training aligned?

**Comment by eric-haibin-lin** [comment]
For better usability / extension, comments from the author of code-R1: 

> From a software engineering perspective, I think it would be more user-friendly for first-time users if we simplify the interface parameters as much as possible! Also, I think improving customization would be very helpful, allowing users to treat verl as a library as much as possible and reduce the need for forking, which would make it easier to accept upstream updates. If there's an interface that allows users to slightly adjust rewards in a prompt group, I think it would enable users to easily customize more algorithm details.
I feel that we can use "prompt: List[response]" as input for the reward function, which would make it easier to adjust the rewards within a prompt group, such as implementing some algorithms for length penalty in Kimi.

**Comment by uygnef** [comment]
> Any key observations to help us identify the root cause underlying this matter would be a significant contribution.
The variances of actor/pg_loss and actor/kl_loss in FSDP and Megatron differ significantly, but average is same.

**Comment by sophiayyya** [comment]
Hi, what does "reference system tuning script for best RL throughput on different types of accelerators" mean? Does it indicate a series of reference scripts containing parallel size, placement methods, etc for different backend? Thanks!

**Comment by BearBiscuit05** [comment]
> Hi, what does "reference system tuning script for best RL throughput on different types of accelerators" mean? Does it indicate a series of reference scripts containing parallel size, placement methods, etc for different backend? Thanks!

Yes, we will first provide the minimum resource requirements for training some models, the optimal configuration requirements, and the corresponding parallel methods, all presented in the form of a table. Later, we may consider improving a script for automatic configuration search.

**Comment by casper-hansen** [comment]
Do you plan to release **Streaming Rollout System (SRS)** to veRL?

![Image](https://github.com/user-attachments/assets/ec1636d4-0324-4cc6-8c63-5cc50ab1b6ae)

**Comment by eric-haibin-lin** [comment]
> Do you plan to release **Streaming Rollout System (SRS)** to veRL?
> 
> ![Image](https://github.com/user-attachments/assets/ec1636d4-0324-4cc6-8c63-5cc50ab1b6ae)

We are looking for contribution from the community since the implementation is based on a custom inference engine not suitable for open source.

**Comment by vadimkantorov** [comment]
For FSDP2, there also remains basic `fsdp_sft_trainer.py` which currently only supports FSDP1 if I understand well:
- https://github.com/volcengine/verl/issues/1504

**Comment by gty1829** [comment]
Do you plan to add audio modality to verl framework？Now I notice that only text，image and video modalities are supported. Thank you

**Comment by Taishi-N324** [comment]
Hi @eric-haibin-lin and the verl maintenance team,

Thank you for all your hard work on verl and for keeping the roadmap transparent.

I would like to ask whether there are plans to add official support for two model families that have recently garnered significant attention:

• Mistral Small (e.g., [Mistral-Small-3.2-24B-Instruct-2506](https://huggingface.co/mistralai/Mistral-Small-3.2-24B-Instruct-2506))
• ERNIE 4.5 (complete collection: https://huggingface.co/collections/baidu/ernie-45-6861cd4c9be84540645f35c9)

Thank you for considering this request!

**Comment by eric-haibin-lin** [comment]
@Taishi-N324 we are open to merge contributions from the community if the changes to integrate these models can be self-contained and modular

**Comment by eric-haibin-lin** [comment]
moving discussion to Q3 roadmap [#2388](https://github.com/volcengine/verl/issues/2388)

**Comment by ccclyu** [comment]
@Taishi-N324  Mistral-Small and ERNIE series both have vision-language models, which are a bit complicated. Considering the Q3 model engine refactor plan, we may consider to implement after the code reactor is done.
